src/CodeStyle/Core/CodeFixes/FixAllContextHelper.cs

```diff
@@ -0,0 +1,134 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Concurrent;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeFixes;
+
+namespace Microsoft.CodeAnalysis.CodeStyle
+{
+    internal static class FixAllContextHelper
```

Comment:
>FixAllContextHelper [](start = 26, length = 19)

Is this being moved from somewhere?

Reply:
I see that majority of code here is duplication of [FixAllContext.DiagnosticProvider](http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/CodeFixes/FixAllOccurrences/FixAllContext.DiagnosticProvider.cs,25), which is a public abstract class. Can we avoid duplicating code and instead sub-type it?

---
In reply to: [222052295](https://github.com/dotnet/roslyn/pull/28873#discussion_r222052295) [](ancestors = 222052295)

Reply:
@mavasani I tried to simplify this but was not successful so far. The required APIs do not appear to have a public entry point anywhere.

Reply:
I meant use this file as a shared file in both projects and then switch the internal helpers in FixAllContext.DiagnosticProvider to call into FixAllContextHelper?

---
In reply to: [222520613](https://github.com/dotnet/roslyn/pull/28873#discussion_r222520613) [](ancestors = 222520613)

Reply:
I ran into a few problems:

* FixAllContext.DiagnosticProvider incorrectly special-cases generated code (can be fixed during the refactoring)
* FixAllContext.ProgressTracker is not a public API (blocks the refactoring)
* The aggregation logic in FixAllContext.DiagnosticProvider is non-deterministic (can be fixed during the refactoring)

Reply:
> FixAllContext.DiagnosticProvider incorrectly special-cases generated code (can be fixed during the refactoring)

Addressed by making `isGeneratedCode` a callback parameter.

> FixAllContext.ProgressTracker is not a public API

Fixed by passing this as a parameter and allowing `null` arguments outside the IDE.

> The aggregation logic in FixAllContext.DiagnosticProvider is non-deterministic

This is an incorrect statement, but I kept the change in behavior for reduced contention when aggregating diagnostics from multiple projects.